rasa/core/nlg/response.py

```diff
@@ -25,28 +25,54 @@ def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
         """
         self.responses = responses
 
+    def _matches_filled_slots(
+        self, filled_slots: Dict[Text, Any], response: Dict[Text, Any],
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get("condition")
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            if filled_slots.get(name) != value:
+                return False
+
+        return True
+
     def _responses_for_utter_action(
-        self, utter_action: Text, output_channel: Text
+        self, utter_action: Text, output_channel: Text, filled_slots: Dict[Text, Any]
     ) -> List[Dict[Text, Any]]:
-        """Return array of responses that fit the channel and action."""
-        channel_responses = []
+        """Returns array of responses that fit the channel, action and condition."""
         default_responses = []
+        conditional_responses = []
 
         for response in self.responses[utter_action]:
-            if response.get("channel") == output_channel:
-                channel_responses.append(response)
-            elif not response.get("channel"):
+            if response.get("condition") is None:
```

Comment:
@ancalita I think this line is causing your bug. You might be adding responses with a specified `channel` to `default_responses`. This in turn might end with you returning channel specific responses in case a not matching `output_channel` is set in line 71.  


```
def test_anca():
    domain = Domain.from_yaml("""
    responses:
        utter_hi:
        - text: "hello"
        - text: "Hello Slack"
          channel: "slack"
    """)

    t = TemplatedNaturalLanguageGenerator(domain.responses)
    responses = t._responses_for_utter_action("utter_hi", "signal", {})

    assert len(responses) == 1
```

Reply:
Thanks, great insight! I have now fixed line 71 to filter out any non-matching channel responses 🥳

Reply:
I think you didn't push yet, right?

Reply:
Nope, not yet 😄 I was going through all the feedback first.